### PR TITLE
[CDAP-15787] Updated JavaScriptTransform for new validation API

### PR DESCRIPTION
**Pull request by dli357**
Updated the JavaScriptTransform.java class for CDAP-15787

### CODE REVIEW

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java [hunk 1]
```diff
@@ -138,7 +139,7 @@ public JavaScriptTransform(Config config) {
   public void configurePipeline(PipelineConfigurer pipelineConfigurer) throws IllegalArgumentException {
     super.configurePipeline(pipelineConfigurer);
     // try evaluating the script to fail application creation if the script is invalid
-    init(null);
+    init(null, pipelineConfigurer.getStageConfigurer().getFailureCollector());
 
     // init intializes schema if present in the config
     Schema outputSchema = (schema == null) ? pipelineConfigurer.getStageConfigurer().getInputSchema() : schema;
```
Comment (CuriousVini) [inline core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java:149]: Not your change, but change schema == null check to Strings.isNullOrEmpty(schema). We should be checking for both null and empty.

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java [hunk 2]
```diff
@@ -346,45 +356,51 @@ private Object decodeSimpleType(Object object, Schema schema) {
...
-  private void init(@Nullable TransformContext context) {
+  private void init(@Nullable TransformContext context, FailureCollector collector) {
     ScriptEngineManager manager = new ScriptEngineManager();
     engine = manager.getEngineByName("JavaScript");
     try {
       engine.eval(ScriptConstants.HELPER_DEFINITION);
     } catch (ScriptException e) {
       // shouldn't happen
-      throw new IllegalStateException("Couldn't define helper functions", e);
+      collector.addFailure("Couldn't define helper functions", "Something went wrong");
```
Comment (CuriousVini) [inline core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java]: keep corrective action null in this case.

Comment (CuriousVini) [reply core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java]: period at the end of error message

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java [hunk 3]
```diff
@@ -394,7 +410,10 @@ private void init(@Nullable TransformContext context) {
     try {
       lookupConfig = GSON.fromJson(config.lookup, LookupConfig.class);
     } catch (JsonSyntaxException e) {
-      throw new IllegalArgumentException("Invalid lookup config. Expected map of string to string", e);
+      collector.addFailure("Invalid lookup config. Expected map of string to string",
+          "Ensure the provided lookup config is valid JSON in the form of a string to string map")
+          .withConfigProperty("lookup");
```
Comment (CuriousVini) [inline core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java]: constant

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java [hunk 4]
```diff
@@ -410,15 +429,20 @@ private void init(@Nullable TransformContext context) {
...
+          "Ensure the input script is valid")
+          .withConfigProperty("script");
     }
     invocable = (Invocable) engine;
     if (config.schema != null) {
       try {
         schema = Schema.parseJson(config.schema);
       } catch (IOException e) {
-        throw new IllegalArgumentException("Unable to parse schema: " + e.getMessage(), e);
+        collector.addFailure(String.format("Unable to parse schema: %s", e.getMessage()),
```
Comment (CuriousVini) [inline core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java]: Invalid schema : %s

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java [hunk 5]
```diff
@@ -196,7 +197,8 @@ void setErrorSchema(Schema schema) {
   public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter) {
     try {
       engine.eval(String.format("var %s = %s;", VARIABLE_NAME, GSON.toJson(input)));
-      Emitter<Map> jsEmitter = new JSEmitter(emitter, schema == null ? input.getSchema() : schema);
+      Emitter<Map> jsEmitter = new JSEmitter(emitter,
```
Comment (CuriousVini) [inline core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java]: we should not be changing transform method. I think the changes in init method should be sufficient.

**Review by CuriousVini — COMMENTED** [review]
is existing test passing?

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java (hunk 1 above)
Comment (dli357) [reply core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java:149]: Would we be able to do this? schema is an object and not a string.

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java (hunk 5 above)
Comment (dli357) [reply core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java]: Fixed

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java (hunk 2 above)
Comment (dli357) [reply core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java]: Fixed

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java (hunk 3 above)
Comment (dli357) [reply core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java]: Fixed

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java (hunk 4 above)
Comment (dli357) [reply core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java]: Fixed

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java (hunk 1 above)
Comment (CuriousVini) [reply core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java:149]: right, this only applies to schema string.

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java [hunk 6]
```diff
@@ -234,8 +238,9 @@ public void emitError(InvalidEntry<Map> invalidEntry) {
     }
 
     public void emitError(Map invalidEntry) {
-      emitter.emitError(getErrorObject(invalidEntry,
-                                       decodeRecord((Map) invalidEntry.get("invalidRecord"), errSchema)));
+      emitter.emitError(
+          getErrorObject(invalidEntry,
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java]: nit: fix indentation:

```
emitter.emitError(
  getErrorObject(invalidEntry,
                 decodeRecord((Map) invalidEntry.get("InvalidRecord"), errSchema)));
```

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java [hunk 7]
```diff
@@ -394,7 +399,10 @@ private void init(@Nullable TransformContext context) {
     try {
       lookupConfig = GSON.fromJson(config.lookup, LookupConfig.class);
     } catch (JsonSyntaxException e) {
-      throw new IllegalArgumentException("Invalid lookup config. Expected map of string to string", e);
+      collector.addFailure("Invalid lookup config.",
+          "Expected JSON map of string to string.")
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java]: nit: indentation:

```
.addFailure("Invalid lookup config."
            "Expected JSON map of string to string.")
  .withConfigProperty(LOOKUP);
```

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java [hunk 8]
```diff
@@ -410,15 +418,19 @@ private void init(@Nullable TransformContext context) {
                                     FUNCTION_NAME, VARIABLE_NAME, EMITTER_NAME, CONTEXT_NAME, config.script);
       engine.eval(script);
     } catch (ScriptException e) {
-      throw new IllegalArgumentException("Invalid script: " + e.getMessage(), e);
+      collector.addFailure(
+          String.format("Invalid script: %s.", e.getMessage()), null)
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java]: indentation

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java [hunk 9]
```diff
@@ -410,15 +418,19 @@ private void init(@Nullable TransformContext context) {
...
+          .withConfigProperty(SCRIPT);
     }
     invocable = (Invocable) engine;
     if (config.schema != null) {
       try {
         schema = Schema.parseJson(config.schema);
       } catch (IOException e) {
-        throw new IllegalArgumentException("Unable to parse schema: " + e.getMessage(), e);
+        collector.addFailure(String.format("Invalid schema: %s.", e.getMessage()),
+            "Output schema must be JSON parseable.").withConfigProperty(SCHEMA);
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java]: indentation

**Review by albertshau — APPROVED** [review]
lgtm after indentation fixes. Please squash commits before merging.

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java (hunk 6 above)
Comment (dli357) [reply core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java]: Fixed

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java (hunk 7 above)
Comment (dli357) [reply core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java]: Fixed

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java (hunk 9 above)
Comment (dli357) [reply core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java]: Fixed

File: core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java (hunk 8 above)
Comment (dli357) [reply core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java]: Fixed

**Pull request merged**